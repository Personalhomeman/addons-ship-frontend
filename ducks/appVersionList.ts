import { Dispatch } from 'redux';
import { createAction, createReducer } from 'deox';

import api from '@/services/ship-api';
import { actionTypeCreator } from '@/utils';
import { AppVersion } from '@/models';
import { RootState } from '@/store';

const $ = actionTypeCreator('APP_VERSION_LIST');

const _fetchAppVersionList = (appSlug: string) => async (dispatch: Dispatch, getState: () => RootState) => {
  dispatch(fetchAppVersionList.next());

  const {
    auth: { token }
  } = getState();
  api.setToken(token);

  try {
    const appVersionList = await api.getAppVersionList(appSlug);

    dispatch(fetchAppVersionList.complete(appVersionList));
  } catch (error) {
    console.error('fetchAppVersionList', error);
    dispatch(fetchAppVersionList.error(error));
  }
};

export const fetchAppVersionList = Object.assign(_fetchAppVersionList, {
  next: createAction($`GET_NEXT`),
  complete: createAction($`GET_COMPLETE`, resolve => (appVersionList: AppVersion[]) => resolve(appVersionList)),
  error: createAction($`GET_ERROR`, resolve => (error: Error) => resolve(error))
});

const defaultState = null as AppVersionListState;

export type AppVersionListState = null | AppVersion[];
export const appVersionList = createReducer(defaultState, handleAction => [
  handleAction(fetchAppVersionList.complete, (_, { payload }) => payload)
]);
