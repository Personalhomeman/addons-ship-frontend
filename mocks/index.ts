import { AppVersion } from '@/models';
import { TestDevice } from '@/models/test-device';

export const mockAppVersion: AppVersion = {
  id: 123,
  version: '1.0.3',
  platform: 'ios',
  buildNumber: 32,
  buildSlug: '13245ads',
  lastUpdate: new Date('2018-09-22T12:42:31Z'),
  description: 'Some semi-long description',
  whatsNew: 'Some longer whats-new',
  minimumOs: '10.2',
  minimumSdk: 'dunno',
  packageName: 'com.package.name',
  certificateExpires: new Date('2018-09-22T12:42:31Z'),
  bundleId: 'bundle-id',
  size: 4567034,
  supportedDeviceTypes: ['iphone', 'ipad'],
  distributionType: 'development',
  iconUrl: 'http://placekitten.com/160/160',
  appName: 'Standup Timer',
  promotionalText: 'Promotional Text',
  keywords: 'Keywords',
  reviewNotes: 'Review Notes',
  supportUrl: 'https://bitrise.io/support',
  marketingUrl: 'https://bitrise.io/marketing',
  scheme: 'piramid',
  configuration: 'canary',
  publicInstallPageURL: 'https://bitrise.io/app/8b334705d8e78276'
};

export const mockTestDevices: TestDevice[] = [
  { deviceId: 'some-device-id-01', deviceType: 'ios', owner: 'test-user-1' },
  { deviceId: 'some-device-id-02', deviceType: 'android', owner: 'test-user-1' },
  { deviceId: 'some-device-id-03', deviceType: 'ios', owner: 'test-user-2' }
];
