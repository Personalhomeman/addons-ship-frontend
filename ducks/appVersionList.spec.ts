jest.mock('@/services/ship-api');
jest.mock('@/utils/file');

import configureMockStore, { MockStoreCreator, MockStoreEnhanced } from 'redux-mock-store';
import thunk from 'redux-thunk';

import { mockAppVersions } from '@/mocks';
import shipApi from '@/services/ship-api';

import { fetchAppVersionList, appVersionList } from './appVersionList';

describe('appVersionList', () => {
  let mockStore: MockStoreCreator, store: MockStoreEnhanced;
  beforeEach(() => {
    mockStore = configureMockStore([thunk.withExtraArgument({ shipApi })]);
    store = mockStore({ auth: { token: 'most-tokenish' } });
  });

  describe('reducer', () => {
    it('loads an app version', () => {
      const state = appVersionList(undefined, fetchAppVersionList.complete(mockAppVersions));

      expect(state).toMatchSnapshot();
    });
  });

  describe('fetchAppVersionList', () => {
    it('fetches an app version', async () => {
      await store.dispatch(fetchAppVersionList('app-slug') as any);

      expect(shipApi.getAppVersionList).toHaveBeenCalledWith('app-slug');

      expect(store.getActions()).toMatchSnapshot();
    });

    it("can't fetch an app version", async () => {
      (shipApi.getAppVersionList as jest.Mock).mockRejectedValueOnce('api had some issue');
      await store.dispatch(fetchAppVersionList('app-slug') as any);

      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
