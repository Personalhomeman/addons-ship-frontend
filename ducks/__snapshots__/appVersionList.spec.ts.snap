// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appVersionList fetchAppVersionList can't fetch an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_LIST_GET_NEXT",
  },
  Object {
    "payload": "api had some issue",
    "type": "APP_VERSION_LIST_GET_ERROR",
  },
]
`;

exports[`appVersionList fetchAppVersionList fetches an app version 1`] = `
Array [
  Object {
    "type": "APP_VERSION_LIST_GET_NEXT",
  },
  Object {
    "type": "APP_VERSION_LIST_GET_COMPLETE",
  },
]
`;

exports[`appVersionList reducer loads an app version 1`] = `
Array [
  Object {
    "appName": "Standup Timer",
    "appSlug": "app-slug-123",
    "buildNumber": 32,
    "buildSlug": "13245ads",
    "bundleId": "bundle-id",
    "certificateExpires": 2018-09-22T12:42:31.000Z,
    "configuration": "canary",
    "description": "Some semi-long description",
    "distributionType": "development",
    "iconUrl": "http://placekitten.com/160/160",
    "id": 123,
    "keywords": "Keywords",
    "lastUpdate": 2018-09-22T12:42:31.000Z,
    "marketingUrl": "https://bitrise.io/marketing",
    "minimumOs": "10.2",
    "minimumSdk": "dunno",
    "packageName": "com.package.name",
    "platform": "ios",
    "promotionalText": "Promotional Text",
    "publicInstallPageURL": "https://bitrise.io/app/8b334705d8e78276",
    "reviewNotes": "Review Notes",
    "scheme": "piramid",
    "size": 4567034,
    "supportUrl": "https://bitrise.io/support",
    "supportedDeviceTypes": Array [
      "iphone",
      "ipad",
    ],
    "version": "1.0.3",
    "whatsNew": "Some longer whats-new",
  },
  Object {
    "appName": "Standup Timer",
    "appSlug": "app-slug-456",
    "buildNumber": 31,
    "buildSlug": "13245ads",
    "bundleId": "bundle-id",
    "certificateExpires": 2018-09-21T12:42:31.000Z,
    "configuration": "canary",
    "description": "Some semi-long description",
    "distributionType": "development",
    "iconUrl": "http://placekitten.com/160/160",
    "id": 456,
    "keywords": "Keywords",
    "lastUpdate": 2018-09-22T12:42:31.000Z,
    "marketingUrl": "https://bitrise.io/marketing",
    "minimumOs": "10.2",
    "minimumSdk": "dunno",
    "packageName": "com.package.name",
    "platform": "ios",
    "promotionalText": "Promotional Text",
    "publicInstallPageURL": "https://bitrise.io/app/8b334705d8e78276",
    "reviewNotes": "Review Notes",
    "scheme": "piramid",
    "size": 4567034,
    "supportUrl": "https://bitrise.io/support",
    "supportedDeviceTypes": Array [
      "iphone",
      "ipad",
    ],
    "version": "1.0.3",
    "whatsNew": "Some longer whats-new",
  },
  Object {
    "appName": "Standup Timer",
    "appSlug": "app-slug-789",
    "buildNumber": 30,
    "buildSlug": "13245ads",
    "bundleId": "bundle-id",
    "certificateExpires": 2018-09-20T12:42:31.000Z,
    "configuration": "canary",
    "description": "Some semi-long description",
    "distributionType": "development",
    "iconUrl": "http://placekitten.com/160/160",
    "id": 789,
    "keywords": "Keywords",
    "lastUpdate": 2018-09-22T12:42:31.000Z,
    "marketingUrl": "https://bitrise.io/marketing",
    "minimumOs": "10.2",
    "minimumSdk": "dunno",
    "packageName": "com.package.name",
    "platform": "ios",
    "promotionalText": "Promotional Text",
    "publicInstallPageURL": "https://bitrise.io/app/8b334705d8e78276",
    "reviewNotes": "Review Notes",
    "scheme": "piramid",
    "size": 4567034,
    "supportUrl": "https://bitrise.io/support",
    "supportedDeviceTypes": Array [
      "iphone",
      "ipad",
    ],
    "version": "1.0.3",
    "whatsNew": "Some longer whats-new",
  },
  Object {
    "appName": "Standup Timer",
    "appSlug": "app-slug-135",
    "buildNumber": 29,
    "buildSlug": "13245ads",
    "bundleId": "bundle-id",
    "certificateExpires": 2018-09-29T12:42:31.000Z,
    "configuration": "canary",
    "description": "Some semi-long description",
    "distributionType": "development",
    "iconUrl": "http://placekitten.com/160/160",
    "id": 135,
    "keywords": "Keywords",
    "lastUpdate": 2018-09-22T12:42:31.000Z,
    "marketingUrl": "https://bitrise.io/marketing",
    "minimumOs": "10.2",
    "minimumSdk": "dunno",
    "packageName": "com.package.name",
    "platform": "ios",
    "promotionalText": "Promotional Text",
    "publicInstallPageURL": "https://bitrise.io/app/8b334705d8e78276",
    "reviewNotes": "Review Notes",
    "scheme": "piramid",
    "size": 4567034,
    "supportUrl": "https://bitrise.io/support",
    "supportedDeviceTypes": Array [
      "iphone",
      "ipad",
    ],
    "version": "1.0.2",
    "whatsNew": "Some longer whats-new",
  },
  Object {
    "appName": "Standup Timer",
    "appSlug": "app-slug-246",
    "buildNumber": 28,
    "buildSlug": "13245ads",
    "bundleId": "bundle-id",
    "certificateExpires": Date { NaN },
    "configuration": "canary",
    "description": "Some semi-long description",
    "distributionType": "development",
    "iconUrl": "http://placekitten.com/160/160",
    "id": 246,
    "keywords": "Keywords",
    "lastUpdate": 2018-09-22T12:42:31.000Z,
    "marketingUrl": "https://bitrise.io/marketing",
    "minimumOs": "10.2",
    "minimumSdk": "dunno",
    "packageName": "com.package.name",
    "platform": "ios",
    "promotionalText": "Promotional Text",
    "publicInstallPageURL": "https://bitrise.io/app/8b334705d8e78276",
    "reviewNotes": "Review Notes",
    "scheme": "piramid",
    "size": 4567034,
    "supportUrl": "https://bitrise.io/support",
    "supportedDeviceTypes": Array [
      "iphone",
      "ipad",
    ],
    "version": "1.0.2",
    "whatsNew": "Some longer whats-new",
  },
]
`;
